File: 우선순위 큐/11279/11279.py
from queue import PriorityQueue
from sys import stdin

n = int(stdin.readline())
q = PriorityQueue()
answer = ''

for i in range(n):
   x = int(stdin.readline())
   if x:
      q.put(-x)
   elif q.empty():
      answer += '0\n'
   else:
      answer += f'{-q.get()}\n'
print(answer)

File: 우선순위 큐/11279/11279.cpp
#include <iostream>
#include <queue>
using namespace std;

int main()
{
   cin.tie(NULL);
   ios::sync_with_stdio(false);
   
   int n, x;
   cin >> n;
   priority_queue<int> q;

   for (int i = 0; i < n; ++i)
   {
      cin >> x;
      if (x)
         q.push(x);
      else if (q.empty())
         cout << 0 << "\n";
      else
      {
         cout << q.top() << "\n";
         q.pop();
      }
   }

   return 0;
}